netplan/cli/commands/apply.py

```diff
@@ -108,7 +108,13 @@ def command_apply(run_generate=True, sync=False, exit_on_error=True):  # pragma:
         # stop backends
         if restart_networkd:
             logging.debug('netplan generated networkd configuration changed, restarting networkd')
-            utils.systemctl_networkd('stop', sync=sync, extra_services=['netplan-wpa@*.service'])
+            utils.systemctl_networkd('stop', sync=sync, extra_services=['netplan-wpa-*.service'])
+            # Historically (up to v0.98) we had netplan-wpa@*.service files, in case of an
+            # upgraded system, we need to make sure to stop those.
+            try:
+                utils.systemctl_networkd('stop', sync=sync, extra_services=['netplan-wpa@*.service'])
```

Comment:
I am wondering a bit about this now. So even though this practically is alright, I don't think I'm a big fan of doing an extra systemctl call just for legacy reasons everytime we run `netplan apply`. I mean, this is fine, but feels ugly. Maybe we could do this in the packaging somehow? I am not a systemd expert, but maybe we could somehow, on package upgrade, symlink the old name in /run/systemd to the new one? This would then be persistent only for one 'run', and would be transparent to the user.

Reply:
You are right and I like the way you are thinking! I'm not an systemd expert either, but after doing some experiments, I think this is unfortunately not possible.

Basically we need to "rename" a running service inside systemd's memory. `netplan generate` will already do the renaming (unlink & re-create) of the legacy & new systemd unit-files for us at reboot or at any `netplan apply` run.

But the problem is, that the old units are still running (can be seen via `systemctl show netplan-wpa@*`), even though their unit-files/symlinks are not existing in the filesystem anymore. Only the systemd process knows the PID of those old, running `netplan-wpa@*` units and can send the proper kill command to stop them.

I was experimenting with `systemctl set-property netplan-wpa@wlan0 Id=netplan-wpa-wlan0 Names=netplan-wpa-wlan0`, trying to rename the running units, but this fails with `Unknown assignment:` errors.

Reply:
That's too bad then, guess we have no choice for now. Was thinking about maybe first querying if there are any services with the old name before running systemctl, but then again it's not a very time wasteful operation.

Reply:
I did some more experiments:

Stopping non-existing services takes approx. 40ms on my machine, while 'is-active' checking takes approx. 10ms. So we should actually opt to only call the stop command, iff they are running (only first 'netplan apply' run after v0.98->v0.99 upgrade).

Also, I verified `systemctl stop` of non-existing services works as expected on older systemd (v237 / bionic) and implemented Dimitri's prefered way of stopping those services, in order to avoid calling `systemctl stop systemd-networkd.service` twice.